src/include/ipc2/header.h

```diff
@@ -0,0 +1,367 @@
+ * same order if used (and usage is determined by status bits in
+ * struct sof_ipc2_hdr).
+ *
+ * The header is designed to support the following use cases.
+ *
+ * 1) Nano messaging via 32bit message and reply. i.e. sending header and
+ *    replying with header only is enough for some use cases like starting and
+ *    stopping global events.
+ *
+ * 2) Micro messaging via 64bit message and reply - send and reply header with
+ *    micro tuple. Expands uses cases from 1) to support stopping and starting
+ *    targeted events.
+ *
+ * 3) Variable size message and reply - like 1) and 2) but messages and replies
+ *    can be variable in size from 32bits upwards. Any use case can be supported
+ *    here since there are no message restrictions.
+ *
+ * 4) Support of legacy ABIs. The header can be prefixed to legacy ABIs by
+ *    using hdr.block = 1 and appending any legacy ABI structure. This allows
+ *    a stable migration path with a small additional word prefixed to
+ *    legacy ABI IPCs.
+ *
+ * 5) High priority messaging. The header now supports a hint for incoming
+ *    message Q handlers so that they can prioritise real time high priority
+ *    messages over standard batch messages. e.g stream start for low latency
+ *    stream could be processed in the Q before sensor config message.
+ *
+ * 6) Datagram mode (no reply needed). The header can tell the message receiver
+ *    that the message does not need to be acknowledged with a reply. Useful
+ *    where the sender may be sending high volume, short lifetime information
+ *    or where the sender does not care about reply (to save cycles on both
+ *    sender and receiver). Door bell protocol would still be followed.
+ *
+ * 7) Message addressing. 32bit sender and receiver addresses can be added in
+ *    header so that messages can be more easily routed to the correct
+ *    destinations. Broadcast messages also supported.
+ *  @{
+ */
+
+/**
+ * Structure Header - Mandatory.
+ * Header metadata for all IPC commands. Identifies IPC message.
+ * Class spelled with K for C++ users.
+ *
+ * @klass: 	Message feature class. e.g. audio, sensor, debug.
+ * @subklass:	Message sub feature. e.g. PCM, kcontrol, Compressed PCM
+ * @action:	Message action. e.g. Start (PCM), Get (kcontrol value)
+ * @ack:	Reply - IPC success, other reply data may or may not follow.
+ */
+struct sof_ipc2_hdr {
+	uint32_t klass : 8;		/**< class ID */
+	uint32_t subklass : 8;		/**< subclass ID */
+	uint32_t action : 8;		/**< action ID */
+	uint32_t ack : 1;		/**< message is a reply - success */
+	uint32_t nack : 1;		/**< message is a reply - fail */
+	uint32_t priority : 1;		/**< 0 normal, 1 high */
+	uint32_t datagram : 1;		/**< is datagram - no reply needed */
+	uint32_t route : 1;		/**< routing data follows */
+	uint32_t size : 1;		/**< size follows */
+	uint32_t elems : 1;		/**< data elems follows */
+	uint32_t block : 1;		/**< data block follows */
```

Comment:
Yes - I did think about use cases where you could have a legacy C struct and some new tuples but that added a bit more complexity. It would be better in this situation (where legacy C structures have to be extended fro new features) to just change to the IPC2 struct for the new feature.
"ack" and "nak" are also mutually exclusive - I will mark these up in the comments.

Reply:
I'm just wondering if maybe we should make those mutually exclusive field n bit enum fields for safety